### DISCUSSION TITLE
Unicorn in PM2 and running programmatically.

### DISCUSSION THREAD

**Discussion by snowy-jaguar** [Kludex/uvicorn#1429] [Q&A]
Hi got a couple questions about Uvicoorn.

I had a look at the deployment tab of the .org website and saw that Uvicoorn can be run from inside a process manager, would process manager 2 work? As that is what I currently use and I am familiar with how it works.

Another question I ahve is about running uvicorn projects programmatically, is `uvicorn.run()` blocking like `asyncio.run()`? I would like to be able to run both my asyncio project and my uvicorn project from the same file as they are related but I can't put `uvicorn.run()` after `asyncio.run()` as the asyncio project blocks the entire process and never 'completes' (intended behaviour).

**Comment by euri10** [comment]
I never used pm2 (actually discovered it) so cant say for sure but I dont see any obvious reason why it could not.

**Reply by snowy-jaguar** [reply]
Okay, thanks for the response. I'm going to keep this open a bit longer in case anyone else knows anything about uvicorn and pm2 and in case someone is able to provide an answer to my second question.

**Reply by euri10** [reply]
sorry didn't answered that, one could use the startup events to run your asyncio project. something like this perhaps:
```
import asyncio

from starlette.applications import Starlette

import uvicorn


async def other_async():
    while True:
        print(1)
        await asyncio.sleep(1)


async def startup():
    task = asyncio.create_task(other_async())


app = Starlette(on_startup=[startup])


if __name__ == "__main__":
    uvicorn.run("two_async:app")
```
